lib/linter/linter.js

```diff
@@ -460,6 +497,38 @@ function normalizeEcmaVersion(parser, ecmaVersion) {
     return ecmaVersion >= 2015 ? ecmaVersion - 2009 : ecmaVersion;
 }
 
+/**
+ * Normalize ECMAScript version from the initial config into languageOptions (year)
+ * format.
+ * @param {number} ecmaVersion ECMAScript version from the initial config
+ * @returns {number} normalized ECMAScript version
+ */
+function normalizeEcmaVersionForLanguageOptions(ecmaVersion) {
+
+    switch (ecmaVersion) {
+        case 3:
+            return 3;
+
+        // void 0 = now ecmaVersion specified so use the default
+        case 5:
+        case void 0:
+            return 5;
+
+        default:
+            if (typeof ecmaVersion === "number") {
+                return ecmaVersion >= 2015 ? ecmaVersion : ecmaVersion + 2009;
+            }
+    }
+
+    /*
+     * We default to the latest supported ecmaVersion for everything else.
+     * Remember, this is for languageOptions.ecmaVersion, which sets the version
+     * that is used for a number of processes inside of ESLint. It's normally
+     * safe to assume people want the latest unless otherwise specified.
+     */
+    return espree.latestEcmaVersion + 2009;
```

Comment:
Since `languageOptions.ecmaVersion` is an ESLint option, and it is used not only for parsing, shouldn't ESLint maintain its own `latestEcmaVersion` instead of using `espree.latestEcmaVersion`?

For example, we could release support for ES2023 in Espree's minor version, which would be then in range of ESLint versions that don't support ES2023. In that case, `languageOptions: { ecmaVersion: "latest" }` in user's configuration will result in `2023`, but as there is no `es2023` environment there will be no ES globals at all.

Reply:
That might be a good idea. Can you open a separate issue to discuss? I don’t want to couple that decision with this PR.

Reply:
>  Can you open a separate issue to discuss?

https://github.com/eslint/eslint/issues/15366